odoo/models.py

```diff
@@ -7455,83 +7539,175 @@ def itemgetter_tuple(items):
+        self.key = ''
 
-def convert_pgerror_unique(model, fields, info, e):
-    # new cursor since we're probably in an error handler in a blown
-    # transaction which may not have been rollbacked/cleaned yet
-    with closing(model.env.registry.cursor()) as cr_tmp:
-        cr_tmp.execute(SQL("""
-            SELECT
-                conname AS "constraint name",
-                t.relname AS "table name",
-                ARRAY(
-                    SELECT attname FROM pg_attribute
-                    WHERE attrelid = conrelid
-                      AND attnum = ANY(conkey)
-                ) as "columns"
-            FROM pg_constraint
-            JOIN pg_class t ON t.oid = conrelid
-            WHERE conname = %s
-        """, e.diag.constraint_name))
-        constraint, table, ufields = cr_tmp.fetchone() or (None, None, None)
-    # if the unique constraint is on an expression or on an other table
-    if not ufields or model._table != table:
-        return {'message': tools.exception_to_unicode(e)}
-
-    # TODO: add stuff from e.diag.message_hint? provides details about the constraint & duplication values but may be localized...
-    if len(ufields) == 1:
-        field_name = ufields[0]
-        field = fields[field_name]
-        message = model.env._(
-            "The value for the field '%(field)s' already exists (this is probably '%(other_field)s' in the current model).",
-            field=field_name,
-            other_field=field['string'],
+    def __set_name__(self, owner, name):
+        assert isinstance(owner, MetaModel)
+        assert name.startswith('_'), "Names of SQL objects in a model must start with '_'"
+        self.key = name[1:]
+        owner._local_sql_objects[self.key] = self
+
+    @property
+    def definition(self) -> str:
+        raise NotImplementedError
+
+    def full_name(self, model: BaseModel) -> str:
+        assert self.key, f"The SQL object is not named ({self.definition})"
+        name = f"{model._table}_{self.key}"
+        return sql.make_identifier(name)
+
+    def _set_message(self, message: ConstraintMessageType):
+        self.message = message
+
+    def get_error_message(self, model: BaseModel, diagnostics=None) -> str:
+        """Build an error message for the object/constraint.
+
+        :param model: Optional model on which the constraint is defined
+        :param diagnostics: Optional diagnostics from the raised exception
+        :return: Translated error for the user
+        """
+        message = self.message
+        if callable(message):
+            return message(model.env, diagnostics)
+        elif message:
+            message = model.env._(message)  # pylint: disable=gettext-variable
+            if diagnostics and (detail := diagnostics.message_detail):
+                detail_label = model.env._('Detail:')
+                message = f"{message}\n{detail_label} {detail}"
+        return message
+
+    def sync_database_object(self, model: BaseModel):
```

Comment:
Sync means that after the call, what is in Odoo is in line with what is in the database. I think it would be more confusing to name it "reflect" because that term is already used in ir.models to reflect into Odoo what is in the database.
I would leave it as is.